source/opt/inline_pass.h

```diff
@@ -0,0 +1,104 @@
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef LIBSPIRV_OPT_INLINE_PASS_H_
+#define LIBSPIRV_OPT_INLINE_PASS_H_
+
+#include <memory>
+#include <vector>
+#include <unordered_map>
+#include <algorithm>
+
+#include "def_use_manager.h"
+#include "module.h"
+#include "pass.h"
+
+namespace spvtools {
+namespace opt {
+
+// See optimizer.hpp for documentation.
+class InlinePass : public Pass {
+ public:
+  InlinePass();
+  const char* name() const override { return "inline"; }
+  Status Process(ir::Module*) override;
+
+ private:
+  // Map from function's result id to function
+  std::unordered_map<uint32_t, ir::Function*> id2function_;
+
+  // Map from block's label id to block
+  std::unordered_map<uint32_t, ir::BasicBlock*> id2block_;
+
+  // Next unused ID
+  uint32_t nextId_;
+
+  // Write the next available Id back to the module
+  inline void FinalizeNextId(ir::Module* module) {
+    module->SetIdBound(nextId_);
+  }
+
+  // Return the next available Id and increment it
+  inline uint32_t TakeNextId() { return nextId_++; }
+
+  // Exhaustively inline all function calls in func as well as in
+  // all code that is inlined into func.
+  bool Inline(ir::Function* func);
+
+  // Find pointer to type in module, return its resultId. 0 if not found.
+  uint32_t FindPointerToType(uint32_t typeId);
+
+  // Add pointer to type to module and return resultId.
+  uint32_t AddPointerToType(uint32_t typeId);
+
+  // Add unconditional branch to labelId to end of block bp
+  void AddBranch(uint32_t labelId, std::unique_ptr<ir::BasicBlock>* bp);
+
+  // Add store of valId to ptrId to end of block bp
+  void AddStore(uint32_t ptrId, uint32_t valId,
+                std::unique_ptr<ir::BasicBlock>* bp);
+
+  // Add load of ptrId into resultId to end of block bp
+  void AddLoad(uint32_t typeId, uint32_t resultId, uint32_t ptrId,
+               std::unique_ptr<ir::BasicBlock>* bp);
+
+  // Return in newBlocks the result of inlining the call at call_ii within
+  // its block call_bi. The block call_bi can just be replaced with the blocks
+  // in newBlocks. Any additional branches are avoided. Debug instructions
+  // are cloned along with their callee instructions. Early returns are
+  // replaced by storing to a local return variable and branching to a 
+  // (created) exit block where the local variable is returned. Formal
+  // parameters are trivially mapped to their actual parameters.
+  //
+  // Also return in newVars additional OpVariable instructions required by 
+  // and to be inserted into the caller function after the block call_bi is 
+  // replaced with newBlocks.
+  void GenInlineCode(std::vector<std::unique_ptr<ir::BasicBlock>>* newBlocks,
```

Comment:
I think there's a tiny kernel that's still true:  The clone of the caller block retains its original ID.  The phi fixups in InlinePass::Inline rely on this, since the phi fixups only occur if newBlocks has at least two blocks.  But this is a pretty minor corner case.